In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os

#set scripts path in variable 'path'
path = r'/Users/constantinmelachrinos/Documents/careerfoundry data analyst/Course files/Data Analytics Immersion/Achievement 4 - InstaCart - Python/Instacart Basket Analysis'

#import Instacart data_sets
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

# Mixed data-types

In [5]:
# make test dataframe for exercise
df_test = pd.DataFrame()
df_test['mix'] = ['a', 'b', 1, True]

In [6]:
#check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [7]:
#convert all data types in column mix to string
df_test['mix'] = df_test['mix'].astype('str')

# Missing values

## Find missing values

In [8]:
#check if there are any missing values in any columns in dataframe products
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
#find missing values in df_prods
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [10]:
#check if it worked
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [11]:
# check how many rows the dataframe has to compare it to the amount of missing values.
df_prods.shape

(49693, 5)

## Addressing missing values

In [12]:
# extract only data with valid product names to new data frame.
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [13]:
# check how many rows it has now.it should be 16 rows less than before.
df_prods_clean.shape

(49677, 5)

## Finding duplicates

In [14]:
# save duplicates in new dataframe called df_dups
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [15]:
#check the duplicates in data frame
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [16]:
# check how many rows it has before deleting duplicates
df_prods_clean.shape

(49677, 5)

## Addressing duplicates

In [17]:
# creat new df with no duplicates and no missing values
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [18]:
#check how many rows have been deleted
df_prods_clean_no_dups.shape

(49672, 5)

## Export cleaned dataframe

In [19]:
#export cleaned products dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'),index=False)


# Tasks

## Step 1
### If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

In [20]:
# run describe funtion to analyze data
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Product_id: there are 49688 product ID's
Aisle_id: there are 134 aisles
Department_id: there are 21 departments
Prices: Minimum price is 1 and max price is 99999 which is odd and needed to be looked at.

## Step 2
### Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [21]:
#run describe function to analyze the data
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Unnamed: its the numbering of the dataset which represent all rows.
Order_id: starts at 1 going till 3421083.
User_id: starts at 1 and going till 206209.
Order_number: has a max value of 100. Could be capped to 100.
orders_day_of_week: Seems alright. Min is 0 which represents monday and max is 6 which represents sunday. The seven days of the week.
order_hour_of_day: seems alright. Min is 0 is representing midnight and max 23 represents hour 23.
days_since_prior_order: min is 0 days and max is 30 days. maybe its capped at 30 days. It has also a different count than all the other columns. Maybe some missing values.

## Step 3
### Check for mixed-type data in your df_ords dataframe.

In [22]:
#check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Apparently there are no mixed data types in df_ords.

## Step 5
### Run a check for missing values in your df_ords dataframe

In [23]:
#check if there are missing values and in which columns they appear
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

as already expected there are missing values in the column 'days_since_prior_order'.
Probably to the fact that new customers have never ordered before. These missing values should maybe replaced with the number 0. But first need to make more analysis on it.

## Step 6
### Address the missing values using an appropriate method.

In [24]:
#locate missing values in df_ords
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [25]:
#check output
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

as we can see the row number is 206209. If our hypothesis is true that these missing data belongs to new customers having placed the first order than we can filter on order_number 1 and it should show the same amount of rows.

In [26]:
df_ords[df_ords['order_number'] == 1]

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

As we can see the row number is the same which means these all belong to first orders and to new customers.
Best approach here in my opinion is to replace the missing values with 0. Since we know that these are their first orders because of order number 1.

In [27]:
#Add new column with name 'first_order' and add value 1 if NaN is true in column days_since_prior_order
df_ords['first_order'] = df_ords['days_since_prior_order'].isnull()*1

In [28]:
#check if replacing of missing values is executed accordingly.
df_ords.head(30)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
1            1   2398795        1             2                   3   
2            2    473747        1             3                   3   
3            3   2254736        1             4                   4   
4            4    431534        1             5                   4   
5            5   3367565        1             6                   2   
6            6    550135        1             7                   1   
7            7   3108588        1             8                   1   
8            8   2295261        1             9                   1   
9            9   2550362        1            10                   4   
10          10   1187899        1            11                   4   
11          11   2168274        2             1                   2   
12          12   1501582        2             2                   5   
13          13   1901567        2             3                   1   
14          14    738281        2             4                   2   
15          15   1673511        2             5                   3   
16          16   1199898        2             6                   2   
17          17   3194192        2             7                   2   
18          18    788338        2             8                   1   
19          19   1718559        2             9                   2   
20          20   1447487        2            10                   1   
21          21   1402090        2            11                   1   
22          22   3186735        2            12                   1   
23          23   3268552        2            13                   4   
24          24    839880        2            14                   3   
25          25   1492625        2            15                   1   
26          26   1374495        3             1                   1   
27          27    444309        3             2                   3   
28          28   3002854        3             3                   3   
29          29   2037211        3             4                   2   

    order_hour_of_day  days_since_prior_order  first_order  
0                   8                     NaN            1  
1                   7                    15.0            0  
2                  12                    21.0            0  
3                   7                    29.0            0  
4                  15                    28.0            0  
5                   7                    19.0            0  
6                   9                    20.0            0  
7                  14                    14.0            0  
8                  16                     0.0            0  
9                   8                    30.0            0  
10                  8                    14.0            0  
11                 11                     NaN            1  
12                 10                    10.0            0  
13                 10                     3.0            0  
14                 10                     8.0            0  
15                 11                     8.0            0  
16                  9                    13.0            0  
17                 12                    14.0            0  
18                 15                    27.0            0  
19                  9                     8.0            0  
20                 11                     6.0            0  
21                 10                    30.0            0  
22                  9                    28.0            0  
23                 11                    30.0            0  
24                 10                    13.0            0  
25                 11                    30.0            0  
26                 14                     NaN            1  
27                 19                     9.0            0  
28                 16            

In [29]:
df_ords=df_ords.drop('Unnamed: 0',axis=1)

In [34]:
df_ords.shape

(3421083, 7)

## Step 7
### check for duplicates in df_ords

In [31]:
# save duplicates in new dataframe called df_dups
df_dups = df_ords[df_ords.duplicated()]

In [32]:
# check if there are any duplicates
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

No duplicates found in df_ords

## Step 9
### Export cleaned dataframe df_ords

In [33]:
#export cleaned products dataframe
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'),index=False)